src/transport/tests/TestSessionManagerDispatch.cpp

```diff
@@ -0,0 +1,305 @@
+/*
+ *
+ *    Copyright (c) 2020-2021 Project CHIP Authors
+ *    All rights reserved.
+ *
+ *    Licensed under the Apache License, Version 2.0 (the "License");
+ *    you may not use this file except in compliance with the License.
+ *    You may obtain a copy of the License at
+ *
+ *        http://www.apache.org/licenses/LICENSE-2.0
+ *
+ *    Unless required by applicable law or agreed to in writing, software
+ *    distributed under the License is distributed on an "AS IS" BASIS,
+ *    WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ *    See the License for the specific language governing permissions and
+ *    limitations under the License.
+ */
+
+/**
+ *    @file
+ *      This file implements unit tests for the SessionManager implementation.
+ */
+
+#define CHIP_ENABLE_TEST_ENCRYPTED_BUFFER_API // Up here in case some other header
+                                              // includes SessionManager.h indirectly
+
+#include <credentials/PersistentStorageOpCertStore.h>
+#include <crypto/PersistentStorageOperationalKeystore.h>
+#include <lib/core/CHIPCore.h>
+#include <lib/support/CodeUtils.h>
+#include <lib/support/TestPersistentStorageDelegate.h>
+#include <lib/support/UnitTestContext.h>
+#include <lib/support/UnitTestRegistration.h>
+#include <protocols/secure_channel/MessageCounterManager.h>
+#include <transport/SessionManager.h>
+#include <transport/TransportMgr.h>
+#include <transport/tests/LoopbackTransportManager.h>
+
+#include <nlbyteorder.h>
+#include <nlunit-test.h>
+
+#include <errno.h>
+
+#undef CHIP_ENABLE_TEST_ENCRYPTED_BUFFER_API
+
+namespace {
+
+using namespace chip;
+using namespace chip::Inet;
+using namespace chip::Transport;
+using namespace chip::Test;
+
+using TestContext = chip::Test::LoopbackTransportManager;
+
+struct MessageTestEntry
+{
+    const char * name;
+
+    const char * peerAddr;
+
+    const char * payload;
+    const char * plain;
+    const char * encrypted;
+    const char * privacy;
+
+    size_t payloadLength;
+    size_t plainLength;
+    size_t encryptedLength;
+    size_t privacyLength;
+
+    const char * encryptKey;
+    const char * privacyKey;
+    const char * epochKey;
+
+    const char * nonce;
+    const char * privacyNonce;
+    const char * compressedFabricId;
+
+    const char * mic;
+
+    uint16_t sessionId;
+    NodeId peerNodeId;
+    FabricIndex fabricIndex;
+};
+
+struct MessageTestEntry theMessageTestVector[] = {
```

Comment:
Can we have more vectors than this and vectors for the behavior of invalid MIC or other negative cases?